### ISSUE TITLE
Can't install from f-droid

### ISSUE THREAD

**Issue by whymzml** [termux/termux-gui#4]
![Screenshot_20211217_045338](https://user-images.githubusercontent.com/90167850/146454759-9d82efa2-5d0f-4dc7-9e80-dfc43dfe4457.jpg)

IDK what's wrong, but I'm just installed all termux app from f-droid

**Comment by tareksander** [comment]
I know what the error is. The app is accidentally signed with the Termux development keys and not with the f-droid keys, so the signature doesn't match with the f-droid version of Termux and it can't install. [I submitted a PR for f-droid to hopefully get that fixed](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/10240).

**Comment by pavan3999** [comment]
same here, downloaded from f-droid.org

**Comment by CAPTAIN1947** [comment]
@whymzml can you share a link?

**Comment by tareksander** [comment]
I checked the signature of the new version on f-droid and it shouldn't work, the signature still doesn't match the one of Termux. You can watch [this PR](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/10240) for progress.

**Comment by rysson** [comment]
@whymzml, please reopen this issue. F-droid 1.0.4 installation stil fails.
It's better keep open to avoid new issues about it IMHO.

@tareksander, thanks for link to PR. And for your work of course.

**Comment by whymzml** [comment]
@tareksander  @rysson  sorry about my decision, for now I installed GitHub version, and when check from f-droid I got this
![Screenshot_20220113_063711](https://user-images.githubusercontent.com/90167850/149538554-224da3d5-2f6c-4c0e-aa9c-962c000fcdb9.jpg)
That's why I close this issue

**Comment by rysson** [comment]
OK, I can install GUI at all. 

F-droid example.
![20220115-002656-595x1310a](https://user-images.githubusercontent.com/5898312/149597989-2c3bfb05-5f19-48b0-8907-2353f6f392ac.png)

**Comment by thelearn-tech** [comment]
> ![Screenshot_20211217_045338](https://user-images.githubusercontent.com/90167850/146454759-9d82efa2-5d0f-4dc7-9e80-dfc43dfe4457.jpg)
> 
> IDK what's wrong, but I'm just installed all termux app from f-droid

Exactly same **Termux:gui** `v0.1.4`

> ![Screenshot_20211217_045338](https://user-images.githubusercontent.com/90167850/146454759-9d82efa2-5d0f-4dc7-9e80-dfc43dfe4457.jpg)
> 
> IDK what's wrong, but I'm just installed all termux app from f-droid

Exactly same case here **termux:gui** `v0.1.4`

**Comment by jasonsyoung** [comment]
Is there a fix for this? I still can't install it.

**Comment by TheRealOne78** [comment]
Same here. ![Screenshot_20220207-114303_F-Droid.png](https://user-images.githubusercontent.com/53640879/152763603-d847c5f9-a914-4460-8b77-8b674f807374.png)

**Comment by plowsof** [comment]
I've just had the same issue. After reading the other issues on this i see that it is because of different versions of the app(s) e.g. termux from fdroid but the termux-gui from github (avoid installing anything from F-Droid to make this work)
Solution: 
- Uninstall everything termux related in F-Droid 
- Download termux from Github (use the universal one for simplicity sake) https://github.com/termux/termux-app/releases/tag/v0.118.0
- Download termux-gui from Github https://github.com/termux/termux-gui/releases

This fixed it for me

**Comment by jasonsyoung** [comment]
This isn't good advice as the F-Droid versions are the only ones being updated from now on. Termux should be removed from the Play Store.

**Comment by Grimler91** [comment]
> This isn't good advice as the F-Droid versions are the only ones being updated from now on. 

I think you misread the previous comment. Both fdroid and github releases are being updated.

> Termux should be removed from the Play Store.

Yes, it should, but none of us who are active here has access to the google account that have control of the old apps

**Comment by indrawanalghifary** [comment]
Cant install from f-droid

**Comment by sqrtrev** [comment]
> I've just had the same issue. After reading the other issues on this i see that it is because of different versions of the app(s) e.g. termux from fdroid but the termux-gui from github (avoid installing anything from F-Droid to make this work) Solution:
> 
> * Uninstall everything termux related in F-Droid
> * Download termux from Github (use the universal one for simplicity sake) https://github.com/termux/termux-app/releases/tag/v0.118.0
> * Download termux-gui from Github https://github.com/termux/termux-gui/releases
> 
> This fixed it for me

This worked on me! Thanks a lot.

**Comment by stedf** [comment]
> I've just had the same issue. After reading the other issues on this i see that it is because of different versions of the app(s) e.g. termux from fdroid but the termux-gui from github (avoid installing anything from F-Droid to make this work) Solution:
> 
>     * Uninstall everything termux related in F-Droid
> 
>     * Download termux from Github (use the universal one for simplicity sake) https://github.com/termux/termux-app/releases/tag/v0.118.0
> 
>     * Download termux-gui from Github https://github.com/termux/termux-gui/releases
> 
> 
> This fixed it for me

The problem persists installing from f-droid. I understand that the above solution is a fix (thank you for that), however that means losing everything already done in termux-app. Can anyone please upload a termux-gui package in f-droid that matches the current f-droid termux-app ?

Thank you in advance.

**Comment by tareksander** [comment]
> however that means losing everything already done in termux-app

You could put all you need in an archive, copy it out of the Termux home folder, reinstall Termux from Github, copy the archive back and unpack it again.

> Can anyone please upload a termux-gui package in f-droid that matches the current f-droid termux-app ?

I can't do that, the people at f-droid have to to that. [I already have a PR open](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/10240), but I'm waiting for a response.

**Comment by Grimler91** [comment]
> You could put all you need in an archive, copy it out of the Termux home folder, reinstall Termux from Github, copy the archive back and unpack it again.

The wiki has steps and suggestions on how to backup: https://wiki.termux.com/wiki/Backing_up_Termux

**Comment by stedf** [comment]
> > however that means losing everything already done in termux-app
> 
> You could put all you need in an archive, copy it out of the Termux home folder, reinstall Termux from Github, copy the archive back and unpack it again.
> 
> > Can anyone please upload a termux-gui package in f-droid that matches the current f-droid termux-app ?
> 
> I can't do that, the people at f-droid have to to that. [I already have a PR open](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/10240), but I'm waiting for a response.



> > You could put all you need in an archive, copy it out of the Termux home folder, reinstall Termux from Github, copy the archive back and unpack it again.
> 
> The wiki has steps and suggestions on how to backup: https://wiki.termux.com/wiki/Backing_up_Termux

thank you both @tareksander and @Grimler91  for the replies and good suggestions. I was trying to avoid the backup and reinstall due to the various symlinks that would be followed or lost.

**Comment by AnasSdiq** [comment]
> I've just had the same issue. After reading the other issues on this i see that it is because of different versions of the app(s) e.g. termux from fdroid but the termux-gui from github (avoid installing anything from F-Droid to make this work)
> Solution: 
> - Uninstall everything termux related in F-Droid 
> - Download termux from Github (use the universal one for simplicity sake) https://github.com/termux/termux-app/releases/tag/v0.118.0
> - Download termux-gui from Github https://github.com/termux/termux-gui/releases
> 
> This fixed it for me 

Still doesn't work for me

**Comment by Techlord210** [comment]
> I've just had the same issue. After reading the other issues on this i see that it is because of different versions of the app(s) e.g. termux from fdroid but the termux-gui from github (avoid installing anything from F-Droid to make this work)
> Solution: 
> - Uninstall everything termux related in F-Droid 
> - Download termux from Github (use the universal one for simplicity sake) https://github.com/termux/termux-app/releases/tag/v0.118.0
> - Download termux-gui from Github https://github.com/termux/termux-gui/releases
> 
> This fixed it for me 


I have installed termux from github but api and style plugin is from f-droid i still cant install

**Comment by tareksander** [comment]
> > I've just had the same issue. After reading the other issues on this i see that it is because of different versions of the app(s) e.g. termux from fdroid but the termux-gui from github (avoid installing anything from F-Droid to make this work)
> > Solution:
> > 
> > * Uninstall everything termux related in F-Droid
> > * Download termux from Github (use the universal one for simplicity sake) https://github.com/termux/termux-app/releases/tag/v0.118.0
> > * Download termux-gui from Github https://github.com/termux/termux-gui/releases
> > 
> > This fixed it for me
> 
> I have installed termux from github but api and style plugin is from f-droid i still cant install

You have to install everything from Github, including Termux:API and Termux:Styling:  
https://github.com/termux/termux-api/releases/tag/v0.50.1  

For Termux:Styling there doesn't seem to be a apk available in the releases, and all the Github Action builds have already expired, so you just have to uninstall it for now.

**Comment by X7md** [comment]
Any progress?

**Comment by tareksander** [comment]
> Any progress?

Not from the f-droid side, but I want to add a plugin system to Termux that doesn't require shared UID, which would also solve the issue: termux/termux-app#2654  
With how long the issue hasn't been solved at f-droid, the new plugin system is probably finished and Termux and Termux:GUI updated to use it before f-droid fixes it.

**Comment by freebrowser1** [comment]
Here not working as well. Then tried with adb:

````
adb install com.termux.gui_5.apk 
Performing Streamed Install
adb: failed to install com.termux.gui_5.apk: Failure [INSTALL_FAILED_SHARED_USER_INCOMPATIBLE: Reconciliation failed...: Reconcile failed: Package com.termux.gui has no signatures that match those in shared user com.termux; ignoring!]
````

**Comment by opera1918** [comment]
It still not working, any suggestion about it?

**Comment by Morphic-droid** [comment]
As of 6/15/22 it STILL won't install!

**Comment by theotheroracle** [comment]
the issue is with fdroid and is slow to resolve ( https://gitlab.com/fdroid/fdroiddata/-/merge_requests/10240 ), the plugin system for termux is being worked on ( https://github.com/termux/termux-app/issues/2654 ) and will also adress the issue when its completed, until then not much can done .

**Comment by Morphic-droid** [comment]
Understood. Thank you

**Comment by Morphic-droid** [comment]
Thank you for your timely response.
________________________________

**Comment by sls1005** [comment]
It still can't be installed. Is it possible to release it under another (package) name, so that it'll be signed with another key?

**Comment by theotheroracle** [comment]
https://github.com/termux/termux-app/pull/2921

is to be merged soon, i would suggest starting to develop the app on these new apis and then fixing the signatures

**Comment by sls1005** [comment]
Okay... But there must be something wrong.
I've downloaded the latest version of termux:gui from f-droid, the apk just built yesterday. And it still can't be intalled, not on a phone with the f-droid version of termux installed. When I tried to extract the signing information, I got:
```txt
X.509, CN=FDroid, OU=FDroid, O=fdroid.org, L=ORG, ST=ORG, C=UK
```
It was signed by F-Droid, but still not with the same key as Termux.

**Comment by tareksander** [comment]
> Okay... But there must be something wrong. I've downloaded the latest version of termux:gui from f-droid, the apk just built yesterday. And it still can't be intalled, not on a phone with the f-droid version of termux installed. When I tried to extract the signing information, I got:
> 
> ```
> X.509, CN=FDroid, OU=FDroid, O=fdroid.org, L=ORG, ST=ORG, C=UK
> ```
> 
> It was signed by F-Droid, but still not with the same key as Termux.

F-Droid has a unique key per app, so that apps cannot access each other's private data. If you want it working on F-Droid sooner, you can also try to [tag the F-Droid team on the merge request](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/10240), maybe they have just forgotten about the issue.

> [termux/termux-app#2921](https://github.com/termux/termux-app/pull/2921)
> 
> is to be merged soon, i would suggest starting to develop the app on these new apis and then fixing the signatures

This issue is actually why I made the new plugin system, but I'll wait for it to be reviewed and merged, in case there are any changes. Adapting it should be very easy anyways. The F-Droid release would still only work after the next Termux update is published on F-Droid, and I don't know when that will happen.

**Comment by Fernal73** [comment]
![Screenshot_20221114_111217](https://user-images.githubusercontent.com/24714685/201584203-5563a2d8-21d9-41c2-9d93-65d1104e8673.jpg)
Is this the same problem?

**Comment by tareksander** [comment]
That seems to be a network or DNS problem, I don't know why it tries to connect to f-droid.org on localhost.

**Comment by Fernal73** [comment]
Output from termux-info:

------
Termux Variables:
TERMUX_APK_RELEASE=F_DROID
TERMUX_APP_PACKAGE_MANAGER=apt
TERMUX_APP_PID=19956
TERMUX_IS_DEBUGGABLE_BUILD=0
TERMUX_MAIN_PACKAGE_FORMAT=debian
TERMUX_VERSION=0.118.0
Packages CPU architecture:
aarch64
Subscribed repositories:
# sources.list
deb https://mirror.bardia.tech/termux/termux-main stable main
# root-repo (sources.list.d/root.list)
deb https://mirror.bardia.tech/termux/termux-root root stable
# x11-repo (sources.list.d/x11.list)
deb https://mirror.bardia.tech/termux/termux-x11 x11 main
Updatable packages:
All packages up to date
termux-tools version:
1.31.0
Android version:
12
Kernel build information:
Linux localhost 5.4.147-qgki-g65a062a09cb7 #1 SMP PREEMPT Fri Oct 21 12:00:10 CST 2022 aarch64 Android
Device manufacturer:
vivo
Device model:
I2126

**Comment by tareksander** [comment]
Can you install other apps from f-droid?

**Comment by Fernal73** [comment]
Apparently not! An f-droid issue then?

**Comment by Fernal73** [comment]
I do have termux installed... Probably installed through f-droid ... some time back.... need to increase the font size...

**Comment by TheRealOne78** [comment]
It's kinda odd, I can't even see termux:gui in fdroid anymore.
Is termux:gui still available in the latest fdroid repo?

**Comment by Fernal73** [comment]
![Screenshot_20221114_155344](https://user-images.githubusercontent.com/24714685/201636460-5d6be2e3-5524-4e1d-8d75-e270e608f70c.jpg)
![Screenshot_20221114_155335](https://user-images.githubusercontent.com/24714685/201636476-b8b50d71-0f22-4d7f-b2ef-076ab2c2f923.jpg)

This shows it. Apk downloaded from website.
https://f-droid.org/

**Comment by TheRealOne78** [comment]
> This shows it. Apk downloaded from website.
> https://f-droid.org/
> 

I am looking at the fdroid website and it just doesn't show termux:gui at all.

![Screenshot_20221114-123153_Fennec.png](https://user-images.githubusercontent.com/53640879/201638336-6f2da0d9-f293-4dc0-b060-5162427f2fea.png)

Did you update the repositories?

**Comment by TheRealOne78** [comment]
I am looking at the gitlab repo of fdroid, and I can't find the `com.termux.gui` dir, except `com.termux.gui.yml`

**Comment by tareksander** [comment]
F-droid reports not having an apk for the app, it seems it got somehow deleted: https://monitor.f-droid.org/builds/nopackages  
Nothing I can do about that.

**Comment by tareksander** [comment]
> I am looking at the gitlab repo of fdroid, and I can't find the `com.termux.gui` dir, except `com.termux.gui.yml`

That is normal, I have the f-droid metadata in my git repo, not on the fdroid-repo, do I don't need to make a pr for every changelog or description change.  
  
Seems like they disabled the build because it didn't install anyways: https://gitlab.com/fdroid/fdroiddata/-/commit/507757a11f5343900f50ba4a592e4045371761a5

**Comment by Fernal73** [comment]
> > This shows it. Apk downloaded from website.
> > https://f-droid.org/
> 
> I am looking at the fdroid website and it just doesn't show termux:gui at all.
> 
> ![Screenshot_20221114-123153_Fennec.png](https://user-images.githubusercontent.com/53640879/201638336-6f2da0d9-f293-4dc0-b060-5162427f2fea.png)
> 
> Did you update the repositories?

F-droid is not installing anything for me with the same error. Should I move this to f-droid repo?

**Comment by Fernal73** [comment]
https://forum.f-droid.org/t/fdroid-is-not-installing-apps-on-android-13/20269 Reported to f-droid

**Comment by khimaros** [comment]
this is still an issue and seems to be blocked by https://gitlab.com/fdroid/admin/-/issues/354

**Comment by licaon-kter** [comment]
disabled for now: https://gitlab.com/fdroid/fdroiddata/-/commit/85803f61c1b67b2f25dc8b93607e9966262952d8

**Comment by khimaros** [comment]
@tareksander mentioned that there was a fix in Termux coming. it may resolve this issue: https://gitlab.com/fdroid/fdroiddata/-/merge_requests/10240#note_1462521781

**Comment by tareksander** [comment]
termux/termux-app#2921 will fix this, as sharedUID won't be needed then to establish a secure connection to the plugin. And when the sharedUID is removed from the manifest, installing from f-droid will work even when the apps are signed with a different key.

**Comment by twaik** [comment]
Actually sharedUid is not really necessary for this. You can start some Java code in apk like it is done in termux-x11. 
i.e. you can do the following:
[0]. starting `TermuxGuiCmdEntryPoint::main()` by `app_process`.
1. Checking if main apk of `termux-gui` is installed.
2. Cheking if it's signature fits to apk executed by `app_process` (comparing with self signature or comparing with string).
3. Sending secure broadcast containing Binder connection (for calling back).
4. Main activity/service receives this broadcast and sends handshake back to process.
5. Connection is established, activity/service requests `ParcelFileDescriptor` of server socket (Unix socket).
6. Process instantiates a socket, wraps it's fd to `ParcelFileDescriptor` and sends it to activity/service.

In this case you need not sharedUid but connection is still safe and clean.

**Comment by tareksander** [comment]
That's possible, but can you use `app_process` as a library? Termux:GUI has another architecture than Termux:X11, there is no server program running in the background, every program connects to the plugin on its own. Also `app_process` has quite a big startup performance hit, I recently updated the python bindings to avoid `am` and use `termux-am` instead if possible, using `app_process` would make that worse again.

**Comment by twaik** [comment]
> That's possible, but can you use `app_process` as a library?

No, but that is not needed. App_process itself will not handle client connections, it is needed only to establish connection with the main code in activity/service and pass there file descriptor of server socket. In the case if you are sure it is enough you can shut down `app_process`.


> Termux:GUI has another architecture than Termux:X11, there is no server program running in the background, every program connects to the plugin on its own

As far as I understand clients connect to Unix socket, is it right? Or you send some intent using termux-am?

> Does Termux:X11 keep the `app_process` open or is the fd just passed to a new instance of xwayland?

It works a bit different way. Main code of termux-x11 is running in app_process, the code running in activity is pretty stupid client which sends input events to server with my own protocol and passes Surface to server when it is changed. It is written to establish new connection when client (activity) considers it is broken or must be renewed. So app_process must work all the time. 

You can start server with termux-services and keep it open. It can send new intent every time you are establishing new connection or when you need it.

**Comment by tareksander** [comment]
The current model is like Termux:API, clients open 2 random sockets in the abstract namespace and send the addresses to the plugin via a broadcast. How does Termux:X11 handle the app process getting killed? Is it automatically restarted?

**Comment by twaik** [comment]
Actually you can make it work with one socket since sockets work like bidirectional pipes.

**Comment by tareksander** [comment]
No, I need 2 sockets: one for the requests and responses, and one for the asynchronous events. Having events mix in with the responses would make the client side unneccessarily complex.

**Comment by twaik** [comment]
X server is contained in the code started with app_process so restarting it is pretty much useless. 

But Activity is stupid client. I intentionally made this to avoid situations when user accidentally closes/kills activity and that kills the whole X session. When it starts it sends signal to `app_process` based server and server sends broadcast (targeted/restricted to `com.termux.x11` package) with Binder instance. When activity receives this broadcast it establishes connection with server (simply requests socket fd obtained by `socketpair`) and passes socket to native part (which is X client working with `libxcb` instead of `libX11`). It is pretty much simple and reliable. Since activity does not have any important parts and does not keep any states it can be restarted easily.

In your case it will be better to make termux-am somehow pass file descriptor to activity/service.
I am advicing somethig like this.
1. termux-am passes commandline **and related fds** to socket.
2. termux-am sends intent containing Binder instance in Bundle. This binder will be used for getting ParcelFileDescriptor-wrapped fd (sending ParcelFileDescriptors itself in Bundle is restricted).
3. Your activity/service receives this intent and sends through Binder a command like `getFd` to get needed file descriptor. 
4. It connects uses file descriptor it gets like in any other case.

**Comment by twaik** [comment]
Ok.

**Comment by khimaros** [comment]
any word on this? I would love to use but don't want to uninstall and lose all of my configuration

**Comment by khimaros** [comment]
see also https://github.com/termux/termux-app/issues/2654

**Comment by ztestgit777** [comment]
F-Droid has removed or disabled the app.
Fix this?

**Comment by agnostic-apollo** [comment]
F-Droid release should be available soon.

- https://gitlab.com/fdroid/admin/-/issues/354#note_2336751805
- https://gitlab.com/fdroid/fdroiddata/-/commit/36ad0c7132a784ea4cda9fcf0e009ab507430bf4

**Comment by licaon-kter** [comment]
> soon

https://gitlab.com/fdroid/wiki/-/wikis/FAQ#how-long-does-it-take-for-my-app-to-show-up-on-website-and-client

**Comment by licaon-kter** [comment]
no, not really, but we've seen the issue and a fix is deployed, it will take 2 cycles to make it right

**Comment by tareksander** [comment]
I think the 2nd build cycle is complete, it installs fine with fdroid Termux now.
![Screenshot_20250221-113912~2.png](https://github.com/user-attachments/assets/4d637070-1de8-4718-a5f8-3eb645116db9)

**Comment by twaik** [comment]
Probably you could avoid messing with signatures by using signature-agnostic design like in termux-x11 :). It is not that hard even counting the fact it should run some JVM code in termux environment (which can be daemonized).

**Comment by DaRandomCube** [comment]
wait it's closed?
after nearly 4 years?
oh my god
i remember when i first complained about it, i forgot what was my first comment

**Comment by DaRandomCube** [comment]
> > I've just had the same issue. After reading the other issues on this i see that it is because of different versions of the app(s) e.g. termux from fdroid but the termux-gui from github (avoid installing anything from F-Droid to make this work)
> > Solution:
> > 
> > * Uninstall everything termux related in F-Droid
> > * Download termux from Github (use the universal one for simplicity sake) https://github.com/termux/termux-app/releases/tag/v0.118.0
> > * Download termux-gui from Github https://github.com/termux/termux-gui/releases
> > 
> > This fixed it for me
> 
> Still doesn't work for me

oh god, here is my first comment

**Comment by microbearsmb** [comment]
No way this actually happened! I have a project that's been waiting on this for a year and a half. Just opened f-droid, and look what's at the top of the list. Thank you so very much  @tareksander and @agnostic-apollo for all the development that made this happen!

**Comment by NoNameWasDefined** [comment]
This is now past ! I don't need anymore the Github version of Termux and each plugins.

**Comment by Manamama** [comment]
As per above: either or. For me: installing the F-Droid https://f-droid.org/packages/com.termux.gui version was needed due to the conflicting keys if trying the one from the Releases here. 
Ref: 

![Image](https://github.com/user-attachments/assets/d59d85ad-6359-4c41-9799-6fcde3f237b1)

The demo apps https://github.com/tareksander/termux-gui-python-bindings/blob/main/tutorial/helloworld.py and the Settings GUI work fine, so thank you.

**Comment by licaon-kter** [comment]
As expected, as the F-Droid version is signed by... F-Droid.
